src/Radicle/Internal/HttpStorage.hs

```diff
@@ -94,15 +94,22 @@ httpStoragePrimFns' mgr =
               \ chain located at the URL for evaluation."
             , \url values -> do
                 res <- liftIO $ runClientM' url mgr (submit $ toList values)
-                pure $ bimap formatServantError (const ()) res
+                pure $ case res of
+                    Left servantError -> Left $ formatServantError servantError
+                    Right val -> case fromRad val of
+                        Left err -> Left $ "cannot parse server response: " <> err
+                        Right n -> Right n
             )
         , storageReceive =
             ( "receive!"
             , "Given a URL (string) and a integral number `n`, queries the remote chain\
               \ for the last `n` inputs that have been evaluated."
-            , \url index -> do
+            , \url maybeIndex -> do
+                let index = fromMaybe (-1) maybeIndex
                 res <- liftIO $ runClientM' url mgr (since index)
-                pure $ first formatServantError res
+                pure $ case res of
+                    Left err     -> Left $ formatServantError err
+                    Right values -> Right (length values + index, values)
```

Comment:
The `length value + index` is depending on implementation details of the Http server. Maybe it would be best to make the Http server's API more similar to a `StorageBackend`, that is, return the index of the last input returned.